File: blackjack/elixir/outcomes.ex

defmodule Outcomes do
  def partitions(cards, subtotal) do
    Enum.sum(
    for i <- 0..9, elem(cards,i)>0 do
      cond do
	subtotal+i+1 > 21 -> 0
	subtotal+i+1==21 -> 1
	subtotal+i+1 < 21 ->
	  1+partitions(put_elem(cards, i, elem(cards,i)-1), subtotal+i+1)
      end
    end
    )
  end
end

deck = Tuple.append(4 |> Tuple.duplicate(9), 16)

d =
  Enum.sum(
    for i <- 0..9 do
      new_deck = put_elem(deck, i, elem(deck, i)-1)
      p =
	Enum.sum(
	  for j <- 0..9 do
	    Outcomes.partitions(put_elem(new_deck, j, elem(new_deck, j)-1), j+1)
	  end
	)
      IO.puts "Dealer showing #{i} partitions = #{p}"
      p
    end
  )

IO.puts "Total partitions = #{d}"

File: blackjack/clojure/outcomes.clj
(ns outcomes
  (:gen-class))

; (use 'clojure.pprint)

(defn partitions [cards subtotal]
  ; (pprint cards)
  ; (println subtotal)
  (def m 0)
   (dotimes [i 10]
     (when (> (aget cards i) 0)
       (def total (+ subtotal (+ i 1)))
       (cond
         (< total 21)
           (do 
             (def m (+ m 1))
             (aset cards i (- (aget cards i) 1))
             (def m (+ m (partitions cards total)))
             (aset cards i (+ (aget cards i) 1))
           )
         (= total 21)
           (def m (+ m 1))
       )
     )
   )
 m)

(def deck (into-array Integer/TYPE [4 4 4 4 4 4 4 4 4 16]))

(def d 0)

(dotimes [i 10]
  ; (pprint deck)
  (aset deck i (- (aget deck i) 1))
  (def p 0)
  (dotimes [j 10]
    (aset deck j (- (aget deck j) 1))
    (def p (+ p (partitions deck (+ j 1))))
    (aset deck j (+ (aget deck j) 1))
  )
  (println (list i p))
  (def d (+ d p))
  (aset deck i (+ (aget deck i) 1))
)

(println d)
